lib/Int.ts

```diff
@@ -0,0 +1,215 @@
+
+const mulFactory = (multiplicand: MetaInteger) => (multiplier: Int): any => {
+   if (!multiplicand.validateSize(multiplier)) {
+       throw new InconsistentSizeError(multiplicand._size, multiplier._size);
+    } else {
+        return resultTyper(multiplicand)(multiplicand._value.multipliedBy(multiplier._value));
+   }
+};
+
+const divFactory = (dividend: MetaInteger) => (divisor: Int): any => {
+   if (!dividend.validateSize(divisor)) {
+       throw new InconsistentSizeError(dividend._size, divisor._size);
+    } else { return pipe (
+            checkDivByZero(dividend._value)(divisor._value),
+            bigNumberOrThrowError,
+            resultTyper(dividend),
+        )(dividend._value.dividedBy(divisor._value));
+   }
+};
+
+// Typers
+const factoryTyper = <U>(uint: MetaInteger) => (factory: (internal: MetaInteger) => (external: Int) => U ) => factory(uint);
+
+const resultTyper = (initial: MetaInteger) => (result: BigNumber) => {
+    if (initial._size === 8) { return Int8(result);
+    } else if (initial._size === 16) { return Int16(result);
+    } else if (initial._size === 32) { return Int32(result);
+    } else if (initial._size === 64) { return Int64(result.toString());
+    } else { throw new TypeNotSupportedError(); }
+};
+
+// Convert Constructor Inputs to BigNumbers
+const emptyValueToZero = (x) => x ? x : 0;
+const notFloat = (x: number): boolean => x % 1 === 0;
+
+const inputTypeToBigNumber = (value?: number | string | BigNumber): BigNumber | Error => {
+    if (value instanceof BigNumber) {
+        return value.isInteger() ? value : new FloatingPointNotSupportedError();
+    } else if (typeof value === "string") {
+        return notFloat(parseInt(value, 10)) ? new BigNumber(value) : new FloatingPointNotSupportedError();
+    } else if (typeof value === "number") {
+        return notFloat(value) ? new BigNumber(value) : new FloatingPointNotSupportedError();
+    }
+};
+
+const sizeCheck = (size: number) => (value: BigNumber | Error): BigNumber | Error => {
+    if (value instanceof Error) { return value; }
+    const numSize = value.toString(2).length;
+    const adjustedSize = value >= new BigNumber(0) ? numSize + 1 : numSize;
+    return adjustedSize <= size ? value : new InvalidSizeError(numSize);
+};
+
+// Int Factory
+const buildMetaInt = (size: number) => (value: BigNumber): MetaInteger => ({
+    _value: value,
+    _size: size,
+    validateSize: validateSize(size),
+});
+
+const signInteger = (metaInt: MetaInteger) => ( {...metaInt, _isPositive: metaInt._value >= new BigNumber(0)} );
+
+const addMathMethods = <T>() => (metaInt: MetaInteger) => ({
+    ...metaInt,
+    add: factoryTyper<T>(metaInt)(addFactory),
+    sub: factoryTyper<T>(metaInt)(subFactory),
+    mul: factoryTyper<T>(metaInt)(mulFactory),
+    div: factoryTyper<T>(metaInt)(divFactory),
+});
+
+const composeObjects = <T>(x) => (y: MetaInteger): T => Object.assign(y, x) as T;
+
+const Int8 = (value?: number | string | BigNumber): Int8 => pipe(
```

Comment:
Would it be possible to abstract these declarations bit further? I'm thinking there is a lot of duplicated code here, would a generatePipe() function be feasible to generalize the flow? (ie. create a function that takes the generic T and size and returns the pipe sequence)